components/cldr-json-data-provider/src/transform/dates.rs

```diff
@@ -0,0 +1,368 @@
+use crate::cldr_langid::CldrLangID;
+use crate::error::Error;
+use crate::reader::{get_subdirectories, open_reader};
+use crate::support::DataKeySupport;
+use crate::CldrPaths;
+use icu_data_provider::iter::DataEntryCollection;
+use icu_data_provider::prelude::*;
+use icu_data_provider::structs::dates::*;
+use std::convert::TryFrom;
+use std::marker::PhantomData;
+
+/// A data provider reading from CLDR JSON dates files.
+#[derive(PartialEq, Debug)]
+pub struct DatesProvider<'d> {
+    data: Vec<(CldrLangID, cldr_json::LangDates)>,
+    _phantom: PhantomData<&'d ()>, // placeholder for when we need the lifetime param
+}
+
+impl TryFrom<&CldrPaths> for DatesProvider<'_> {
+    type Error = Error;
+    fn try_from(cldr_paths: &CldrPaths) -> Result<Self, Self::Error> {
+        let mut data = vec![];
+
+        let path = cldr_paths.cldr_dates.clone()?.join("main");
+
+        let locale_dirs = get_subdirectories(&path).unwrap();
+
+        for dir in locale_dirs {
+            let path = dir.join("ca-gregorian.json");
+
+            let langid: CldrLangID = dir.file_name().unwrap().to_str().unwrap().parse().unwrap();
+            let resource: cldr_json::Resource =
+                serde_json::from_reader(open_reader(path)?).unwrap();
+            let lang_dates = resource
+                .main
+                .0
+                .into_iter()
+                .find(|(lid, _)| lid == &langid)
+                .map(|(_, val)| val)
+                .unwrap();
+            match data.binary_search_by_key(&&langid, |(langid, _)| langid) {
+                Ok(_) => panic!("Duplicated entry"),
+                Err(idx) => {
+                    data.insert(idx, (langid, lang_dates));
+                }
+            }
+        }
+
+        Ok(Self {
+            data,
+            _phantom: PhantomData,
+        })
+    }
+}
+
+impl<'d> DataKeySupport for DatesProvider<'d> {
+    fn supports_key(data_key: &DataKey) -> Result<(), DataError> {
+        if data_key.category != DataCategory::Dates {
+            return Err((&data_key.category).into());
+        }
+        if data_key.version != 1 {
+            return Err(data_key.into());
+        }
+        Ok(())
+    }
+}
+
+impl<'d> DatesProvider<'d> {
+    fn get_dates_for(
```

Comment:
I don't think you need this function in DatesProvider.  It is implemented in PluralsProvider only to resolve the cardinal/ordinal fork.  There is no such fork in this provider.

You can move the `binary_search_by_key` into the call site.

Reply:
I actually like this separation for readability. Are you ok with leaving it?

Reply:
Okay, sure.  The contract of the function is different between PluralsProvider and DatesProvider, but that's not really a problem.